### DISCUSSION TITLE
When will gitblit 1.10.0 be released?

### DISCUSSION THREAD

**Discussion by flaix** [gitblit/gitblit#1440] [General]
Do you need help? Anything I can do?

_Originally posted by @andrm in https://github.com/gitblit/gitblit/issues/1419#issuecomment-1338934485_

**Comment by flaix** [comment]
Well, that probably depends on what you are interested in.
There are two minor issues remaining to be fixed or verified. One is to verify the problem in #1423 and that PR #1425 fixed it.
Other than that, there are two more issued in the milestone for 1.10.0 which I would now move to the 1.10.1 milestone, in order to not keep piling on more things.
The last one is the rather large topic of metrics that I don't know how to split up for help right now.

Other than that, if you are running a Gitblit instance, you could run the latest nightly build to see if it is really stable enough to be released. If you use tickets, though, you shouldn't because pushing to tickets seems to have been broken.
If you run Gitblit on Windows even better, try the latest to find issues that need fixing before release, as I ma not running it on Windows anymore. Stuff like that.

**Comment by andrm** [comment]
SNI: Maybe solved by a jgit upgrade. 
SSHD: Maybe fixed in apache sshd impl.

I will try to run the latest and see if it works.

**Comment by andrm** [comment]
1.10.0 snapshot looks good to me. ed25519 seems to be preferred for now. The old workarounds 
 HostKeyAlgorithms      +ssh-rsa
 PubkeyAcceptedKeyTypes +ssh-rsa
strill work. But if you choose to omit them, the ssh client will complain about changed host key (because it will now use the ssh-ed25519 )

**Comment by flaix** [comment]
Interesting find, thank you. I will have to try the upgrade in order to find some wording to put into the release notes. Didn't think of that. I wonder why I didn't not encounter that. I didn't think it would complain about a changed host key but add a new one.

**Comment by andrm** [comment]
Can you check this code in SshDaemon again:

```
		// Generate host RSA and DSA keypairs and create the host keypair provider
		File rsaKeyStore = new File(gitblit.getBaseFolder(), "ssh-rsa-hostkey.pem");
		File dsaKeyStore = new File(gitblit.getBaseFolder(), "ssh-dsa-hostkey.pem");
		File ecdsaKeyStore = new File(gitblit.getBaseFolder(), "ssh-ecdsa-hostkey.pem");
		File eddsaKeyStore = new File(gitblit.getBaseFolder(), "ssh-eddsa-hostkey.pem");
		File ed25519KeyStore = new File(gitblit.getBaseFolder(), "ssh-ed25519-hostkey.pem");
		generateKeyPair(rsaKeyStore, "RSA", 2048);
		generateKeyPair(ecdsaKeyStore, "ECDSA", 256);
		generateKeyPair(eddsaKeyStore, "EdDSA", 0);
		FileKeyPairProvider hostKeyPairProvider = new FileKeyPairProvider();
		hostKeyPairProvider.setFiles(new String [] { ecdsaKeyStore.getPath(), eddsaKeyStore.getPath(), ed25519KeyStore.getPath(), rsaKeyStore.getPath(), dsaKeyStore.getPath() });
```
The file ssh-ed25519-hostkey.pem doesn't show up in my gitbit base dir, but the ssh client has this type saved in known_hosts.
The other files are there, rsa and dsa were left untouched, ssh-ecdsa-hostkey.pem and ssh-eddsa-hostkey.pem were created.

**Comment by flaix** [comment]
This looks weird if you look at the code, yes, but is actually intentional.

The story here is that this version of SSHD requires a certain third-party library for Ed25519 keys. Probably because BouncyCastle didn't have support yet at the time. I do hope that the future versions will be able to simply use BouncyCastle so that I can get rid of that extra library.

The keys that that library creates and writes are not in a proper OpenSSH format that other tools would recognise. In turn, it cannot handle proper OpenSSH Ed25519 created keys. Which is why there is special code already created accounting for both types, hoping to be able to support the normal type in the future.
Ed25519 keys are essentially EdDSA keys. The library currently in play also doesn't use the Ed25519 algorithm name but uses EdDSA, but the generated keys are actually Ed25519 keys. That is why the keys generated with the current library are stored as "ssh-eddsa-hostkey" but show up in your client as ed25519 keys. This is so that after (hopefully) switching to BouncyCastle only, then the "ssh-ed25519-hostkey" filename can be used.

In principal, you should be able to generate a Ed25519 keypair with OpenSSH, store it there under the "ssh-ed25519-hostkey" name and Gitblit should be able to pick it up and use.

**Comment by andrm** [comment]
Thanks. Maybe add this as comment to the code.

**Comment by flaix** [comment]
Since version 1.10.0 has been released, I am setting this topic to closed.
